specs/_features/eip7251/beacon-chain.md

```diff
@@ -0,0 +1,827 @@
+        has_execution_withdrawal_credential(validator)
+        # Verify withdrawal credentials
+        and validator.withdrawal_credentials[12:] == execution_layer_withdraw_request.source_address
+        # Verify the validator is active
+        and is_active_validator(validator, get_current_epoch(state))
+        # Verify exit has not been initiated, and slashed
+        and validator.exit_epoch == FAR_FUTURE_EPOCH
+        # Verify the validator has been active long enough
+        and get_current_epoch(state) >= validator.activation_epoch + SHARD_COMMITTEE_PERIOD
+    ):
+        return
+    # New condition: only allow partial withdrawals with compounding withdrawal credentials
+    if not (is_full_exit_request or has_compounding_withdrawal_credential(validator)):
+        return
+
+    pending_balance_to_withdraw = sum(
+        item.amount for item in state.pending_partial_withdrawals if item.index == index
+    )
+    # only exit validator if it has no pending withdrawals in the queue
+    if is_full_exit_request and pending_balance_to_withdraw == 0:
+        initiate_validator_exit(state, index)
+    elif state.balances[index] > MIN_ACTIVATION_BALANCE + pending_balance_to_withdraw:
+        to_withdraw = min(
+            state.balances[index] - MIN_ACTIVATION_BALANCE - pending_balance_to_withdraw,
+            amount
+        )
+        exit_queue_epoch = compute_exit_epoch_and_update_churn(state, to_withdraw)
+        withdrawable_epoch = Epoch(exit_queue_epoch + MIN_VALIDATOR_WITHDRAWABILITY_DELAY)
+        state.pending_partial_withdrawals.append(PartialWithdrawal(
+            index=index,
+            amount=to_withdraw,
+            withdrawable_epoch=withdrawable_epoch,
+        ))
+```
+
+###### Updated `get_expected_withdrawals`
+
+```python
+def get_expected_withdrawals(state: BeaconState) -> Sequence[Withdrawal]:
+    epoch = get_current_epoch(state)
+    withdrawal_index = state.next_withdrawal_index
+    validator_index = state.next_withdrawal_validator_index
+    withdrawals: List[Withdrawal] = []
+    consumed = 0
+    for withdrawal in state.pending_partial_withdrawals:
+        if withdrawal.withdrawable_epoch > epoch or len(withdrawals) == MAX_WITHDRAWALS_PER_PAYLOAD // 2:
+            break
+
+        validator = state.validators[withdrawal.index]
+        if validator.exit_epoch == FAR_FUTURE_EPOCH and state.balances[withdrawal.index] > MIN_ACTIVATION_BALANCE:
+            withdrawable_balance = min(state.balances[withdrawal.index] - MIN_ACTIVATION_BALANCE, withdrawal.amount)
+            withdrawals.append(Withdrawal(
+                index=withdrawal_index,
+                validator_index=withdrawal.index,
+                address=ExecutionAddress(validator.withdrawal_credentials[12:]),
+                amount=withdrawable_balance,
+            ))
+            withdrawal_index += WithdrawalIndex(1)
+            consumed += 1
+
+    state.pending_partial_withdrawals = state.pending_partial_withdrawals[consumed:] 
```

Comment:
This modifies the state in a "getter" function. That update should happen elsewhere.